### PR TITLE
Improve documentation and warning message for creation of a tensor with from_numpy()

### LINKED ISSUE #46741: Tensor with read only memory mapped numpy array causes seg fault.
## 🐛 Bug

If you use a memory mapped numpy array and open it read only, pytorch won't respect that it's read only. If you accidentally write to it, it causes a seg fault. It does give a warning that you may overwrite your data unexpectedly, but not that you may end up segfaulting.
## To Reproduce

>>> import numpy as np
>>> import torch

>>> A = np.zeros((3,4))
>>> np.save('temp.npy', A)
>>> del A

>>> A = np.load('temp.npy', mmap_mode='r')
>>>  T = torch.from_numpy(A)
>>> T[0,0] = 1

cc @ezyang @gchanan @zou3519 @bdhirsh @heitorschueroff @jlin27 @mruberry @rgommers

### PR DESCRIPTION
**Pull request by leonvol**
Implements very simple changes suggested in the short discussion of the issue. Updated documentation to inform user that creation of tensor with memory mapped read only numpy arrays will probably cause a crash of the program. The displayed warning message was also updated to contain the information about issues concerning the use of a memory mapped read only numpy array. Closes #46741.

### CODE REVIEW

**Comment by mrshenli** [comment]
Hey @leonvol, could you please update the PR title and summary to be more informative? Thanks!

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -3035,8 +3035,10 @@ def merge_dicts(*dicts):
 Creates a :class:`Tensor` from a :class:`numpy.ndarray`.
 
 The returned tensor and :attr:`ndarray` share the same memory. Modifications to
-the tensor will be reflected in the :attr:`ndarray` and vice versa. The returned
-tensor is not resizable.
+the tensor will be reflected in the :attr:`ndarray` and vice versa.
```
Comment (mruberry) [inline torch/_torch_docs.py:3038]: Don't change this paragraph, instead add a warning after the last paragraph and before the example. The warning should say:

"Writing to a tensor created from a read-only NumPy array is not supported and will result in undefined behavior."

File: torch/csrc/utils/tensor_numpy.cpp [hunk 2]
```diff
@@ -136,11 +136,12 @@ at::Tensor tensor_from_numpy(PyObject* obj, bool warn_if_not_writeable/*=true*/)
   if (!PyArray_ISWRITEABLE(array) && warn_if_not_writeable) {
     TORCH_WARN_ONCE(
       "The given NumPy array is not writeable, and PyTorch does "
-      "not support non-writeable tensors. This means you can write to the "
-      "underlying (supposedly non-writeable) NumPy array using the tensor. "
-      "You may want to copy the array to protect its data or make it writeable "
-      "before converting it to a tensor. This type of warning will be "
-      "suppressed for the rest of this program.");
+      "not support non-writeable tensors. This means that writing to the "
```
Comment (mruberry) [inline torch/csrc/utils/tensor_numpy.cpp:139]: Replace the sentence

"This means that writing to the underlying (supposedly non-writeable) NumPy array using the tensor probably result in a crash of the program."

With

"Writing to this tensor is not supported and will result in undefined behavior."

**Comment by mruberry** [comment]
Hey @leonvol! Thanks for this PR. I suggested a couple clarifications for consistency that should be simple to apply.

File: torch/csrc/utils/tensor_numpy.cpp (hunk 2 above)
Comment (leonvol) [reply torch/csrc/utils/tensor_numpy.cpp:139]: Thank you for the feedback. I will make these suggested changes and push immediately.

File: torch/csrc/utils/tensor_numpy.cpp [hunk 3]
```diff
@@ -136,11 +136,12 @@ at::Tensor tensor_from_numpy(PyObject* obj, bool warn_if_not_writeable/*=true*/)
...
     TORCH_WARN_ONCE(
       "The given NumPy array is not writeable, and PyTorch does "
-      "not support non-writeable tensors. This means you can write to the "
-      "underlying (supposedly non-writeable) NumPy array using the tensor. "
-      "You may want to copy the array to protect its data or make it writeable "
-      "before converting it to a tensor. This type of warning will be "
-      "suppressed for the rest of this program.");
+      "not support non-writeable tensors. Writing to this tensor is not "
+      "supported and will result in undefined behavior. "
+      " You may want to copy the "
```
Comment (mruberry) [inline torch/csrc/utils/tensor_numpy.cpp:141]: This sentence would have two spaces after the period. Just delete the space at the start of the quotation on this line.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -3035,8 +3035,10 @@ def merge_dicts(*dicts):
 Creates a :class:`Tensor` from a :class:`numpy.ndarray`.
 
 The returned tensor and :attr:`ndarray` share the same memory. Modifications to
-the tensor will be reflected in the :attr:`ndarray` and vice versa. The returned
-tensor is not resizable.
+the tensor will be reflected in the :attr:`ndarray` and vice versa. 
+Writing to a tensor created from a read-only NumPy array is not supported and
```
Comment (mruberry) [inline torch/_torch_docs.py:3039]: This should be a warning, see here:

https://github.com/pytorch/pytorch/blob/5c3788d5d76f64f6708e0b79f40b1cf45276625a/torch/_torch_docs.py#L339

And it should appear just before the example.

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -3034,15 +3034,17 @@ def merge_dicts(*dicts):
...
-
 It currently accepts :attr:`ndarray` with dtypes of ``numpy.float64``,
 ``numpy.float32``, ``numpy.float16``, ``numpy.complex64``, ``numpy.complex128``,
 ``numpy.int64``, ``numpy.int32``, ``numpy.int16``, ``numpy.int8``, ``numpy.uint8``,
 and ``numpy.bool``.
 
+.. warning::
+    The returned tensor and :attr:`ndarray` share the same memory. Modifications to
+    the tensor will be reflected in the :attr:`ndarray` and vice versa. 
+    Writing to a tensor created from a read-only NumPy array is not supported and
```
Comment (mruberry) [inline torch/_torch_docs.py:3045]: Sorry what I said before was confusing. I meant just this one sentence:

"Writing to a tensor created from a read-only NumPy array is not supported and will result in undefined behavior." 

should be a warning, not the entire paragraph. The rest of the paragraph can go back where it was.

**Review by mruberry — APPROVED** [review]
Thanks @leonvol!

**Comment by mruberry** [comment]
Had to unland this, looks like it acquired a merge conflict that affected the doc formatting and is now throwing an error:

```
Jan 06 00:10:52 /opt/conda/lib/python3.6/site-packages/torch/__init__.py:docstring of torch.from_numpy:15: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

@leonvol, would you try rebasing this and resolve the doc build complaint?

**Comment by leonvol** [comment]
Is this fixed now? @mruberry @hwangdeyu

**Comment by mruberry** [comment]
> Is this fixed now? @mruberry @hwangdeyu

I think you still need to rebase this and fix the doc build failure, right?

**Comment by leonvol** [comment]
Can someone please take over this PR? I cannot build the docs locally and have issues rebasing.

**Comment by SmrutiSikha** [comment]
@leonvol @mruberry  I want to take over this PR? But can you help me to setup this locally? Can you tell me what are the problems that you faced?

**Comment by mruberry** [comment]
Hi @SmrutiSikha, thanks for your interest. The actual change is pretty simple. It's currently buried under a ton of unrelated commits, however. Find the original change and create a new PR with just it.

**Pull request closed without merging**